pex/resolver.py

```diff
@@ -167,420 +39,298 @@ def __new__(cls, requirement, distribution):
-  all compatible distributions encountered for each requirement rather than the single best
-  distribution encountered for each requirement.  This prevents situations where ``tornado`` and
-  ``tornado==2.0`` could be treated as incompatible with each other because the "best
-  distribution" when encountering ``tornado`` was tornado 3.0.  Instead, ``resolve`` maintains the
-  set of compatible distributions for each requirement as it is encountered, and iteratively filters
-  the set.  If the set of distributions ever becomes empty, then ``Unsatisfiable`` is raised.
-
-  .. versionchanged:: 0.8
-    A number of keywords were added to make requirement resolution slightly easier to configure.
-    The optional ``obtainer`` keyword was replaced by ``fetchers``, ``translator``, ``context``,
-    ``threads``, ``precedence``, ``cache`` and ``cache_ttl``, also all optional keywords.
-
-  .. versionchanged:: 1.0
-    The ``translator`` and ``threads`` keywords have been removed.  The choice of threading
-    policy is now implicit.  The choice of translation policy is dictated by ``precedence``
-    directly.
-
-  .. versionchanged:: 1.0
-    ``resolver`` is now just a wrapper around the :class:`Resolver` and :class:`CachingResolver`
-    classes.
-
-  .. versionchanged:: 1.5.0
-    The ``pkg_blacklist``  has been removed and the return type changed to a list of
-    :class:`ResolvedDistribution`.
   """
 
-  builder = ResolverOptionsBuilder(fetchers=fetchers,
-                                   allow_prereleases=allow_prereleases,
-                                   use_manylinux=use_manylinux,
-                                   transitive=transitive,
-                                   precedence=precedence,
-                                   context=context)
-
-  if cache:
-    resolver = CachingResolver(cache,
-                               cache_ttl,
-                               allow_prereleases=allow_prereleases,
-                               use_manylinux=use_manylinux,
-                               transitive=transitive,
-                               interpreter=interpreter,
-                               platform=platform)
-  else:
-    resolver = Resolver(allow_prereleases=allow_prereleases,
-                        use_manylinux=use_manylinux,
-                        transitive=transitive,
-                        interpreter=interpreter,
-                        platform=platform)
-
-  return resolver.resolve(resolvables_from_iterable(requirements, builder, interpreter=interpreter))
-
-
-def resolve_multi(requirements,
-                  fetchers=None,
+  if not requirements and not requirement_files:
+    # Nothing to resolve.
+    return []
+
+  workspace = safe_mkdtemp()
+  resolved_dists = os.path.join(workspace, 'resolved')
+  built_wheels = os.path.join(workspace, 'wheels')
+  installed_chroots = os.path.join(workspace, 'chroots')
+
+  # 1. Resolve
```

Comment:
I meant to add a comment that layed out the reasons for the # 1. Resolve, # 2. Build, # 3. Install comments seen below. I can do so if desired, but here it is:

You'd think we might be able to just `pip install` all the requirements, but pexes can be multiplatform / multiinterpreter in which case only a subset of distributions resolved into the PEX should be activated. Sometimes there are platform specific wheels, sometimes python version specific dists (backports being the common case). As such, we need to be able to add each resolved distribution to the `sys.path` individually (PEXEnvironment handles this selective activation at runtime). Since `pip install` only accepts a single location to install all resolved dists, that won't work. This means we need to seperately resolve all distributions, then install each in their own chroot. To do this we use `pip download` for the resolve and download of all needed distributions. Then `pip install` to install each distribution individually. As a complicating factor, some requirements are local; ie `./path/to/setup-py/project`, in which case no wheel or sdist is downloaded even though transitive dependencies of the local project are. As such we insert a `pip wheel` step to build wheels from any local project requirements found.

Reply:
Done.